src/hyp_map.rs

```diff
@@ -96,8 +96,44 @@ impl HypMapFixedRegion {
     }
 }
 
-/// Represents a virtual address region that must be allocated and populated to be mapped.
-struct HypMapPopulatedRegion {
+// Represents an area that must be reset to restore U-mode original state.
+struct UmodeResetArea {
+    // The physical address where this region starts.
+    paddr: PageAddr<SupervisorPhys>,
+    // The size of the memory mapped for this region. A PageSize::Size4k multiple, bigger than `size`.
+    mapped_size: u64,
+    // The size in bytes of this region.
+    size: usize,
+    // Data to be populated at the beginning of the area.
+    data: Option<&'static [u8]>,
+}
+
+impl UmodeResetArea {
+    fn reset(&self) {
+        let dest = self.paddr.bits() as *mut u8;
+        // Clear memory first.
+        // Safety: `dest` has `mapped_size` of memory mapped. The memory is owned by the hypervisor, by
+        // construction.
```

Comment:
either `reset` needs to be unsafe, or a `UmodeResetArea::new` needs to be added as unsafe and creation of this struct must be restricted to that function. Adding a new is cleaner as it better contains the unsafe code within one location of the user.

As is, safe code can create a `UmodeResetArea` with any pointer, then call reset to scribble on it.

It might require this struct to be its own rs file.

Reply:
Changed in current code to use VA (see `umode_mem.rs`). The structure is created only if the address is in VA area. And the caller must guarantee the U-mode area is mapped (page tables are static, so you need to call it from the right place). In case it isn't, though, what we get is a page fault, rather than memory corruption.